pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -162,36 +244,52 @@ func (e *DockerRegistryServiceController) handleLocationChange(serviceLocation s
 	for _, dockercfgSecret := range dockercfgSecrets {
 		dockercfg := &credentialprovider.DockerConfig{}
 		if err := json.Unmarshal(dockercfgSecret.Data[api.DockerConfigKey], dockercfg); err != nil {
+			// we can't automatically recover this
 			util.HandleError(err)
 			continue
 		}
 
 		dockercfgMap := map[string]credentialprovider.DockerConfigEntry(*dockercfg)
 		keys := sets.KeySet(reflect.ValueOf(dockercfgMap))
-		if len(keys) != 1 {
+		if len(keys) == 0 {
+			// we can't automatically recover this
 			util.HandleError(err)
 			continue
 		}
-		oldKey := keys.List()[0]
-
-		// if there's no change, skip
-		if oldKey == serviceLocation {
+		// if the keys haven't changed, don't make an update and check the next one
+		if keys.Equal(secretKeys) {
 			continue
 		}
 
-		dockercfgMap[serviceLocation] = dockercfgMap[oldKey]
-		delete(dockercfgMap, oldKey)
-		t := credentialprovider.DockerConfig(dockercfgMap)
+		// all keys should have the same value, just take the first-ish one
+		dockerCredentials := dockercfgMap[keys.List()[0]]
+		newDockercfgMap := map[string]credentialprovider.DockerConfigEntry{}
+
+		for key := range secretKeys {
+			newDockercfgMap[key] = dockerCredentials
+		}
+
+		t := credentialprovider.DockerConfig(newDockercfgMap)
 		dockercfg = &t
 
 		dockercfgContent, err := json.Marshal(dockercfg)
 		if err != nil {
+			// we can't automatically recover this either, something went weirdly wrong
 			util.HandleError(err)
 			continue
 		}
 		dockercfgSecret.Data[api.DockerConfigKey] = dockercfgContent
 
-		if _, err := e.client.Secrets(dockercfgSecret.Namespace).Update(dockercfgSecret); err != nil {
+		err = client.RetryOnConflict(client.DefaultBackoff, func() error {
+			_, err := e.client.Secrets(dockercfgSecret.Namespace).Update(dockercfgSecret)
```

Comment:
If you don't update the input, would you ever get a successful update after a conflict? I'd expect more to be inside the retry func